hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -129,6 +129,11 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
     Optional<Long> lockId = Optional.empty();
     try {
       lockId = Optional.of(acquireLock());
+
+      if (base != current()) {
+        throw new CommitFailedException("Cannot commit: stale table metadata for %s.%s", database, tableName);
```

Comment:
This works to ensure that `base.location()` is the same as `currentMetadataLocation()`, but I think that line 155/160 should use `base.location()`. That way the check is for `base` and the current location in the Hive Metastore and doesn't rely on additional state in the `TableOperations` object and a secondary check here. This is a good check to fail fast, but the real check should be the base location and the current HMS location.

Reply:
Won't `base.location()` give us the table location and not the location of the valid metadata file? The valid metadata file is stored in table properties, which we fetch in 154/59.

Reply:
I think `base.location()` will always give us the same value.

Reply:
Do you mean `base.file().location()`?

Reply:
Yes, I meant `base.file().location()`.

Reply:
I think it should be safe to assume `base.file()` is never null if `base` is not null.

Reply:
Agreed.